types/module/simulation.go

```diff
@@ -1,21 +1,44 @@
 package module
 
 import (
+	"encoding/json"
+	"math/rand"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/simulation"
 )
 
+// AppModuleSimulation defines the standard functions that every module should expose
+// for the SDK blockchain simulator
+type AppModuleSimulation interface {
+	// register a func to decode the each module's defined types from their corresponding store key
+	RegisterStoreDecoder(sdk.StoreDecoderRegistry)
+
+	// randomized genesis states
+	GenerateGenesisState(input *SimulationState)
+
+	// randomized module parameters for param change proposals
+	RandomizedParams(r *rand.Rand) []simulation.ParamChange
+}
+
 // SimulationManager defines a simulation manager that provides the high level utility
 // for managing and executing simulation functionalities for a group of modules
 type SimulationManager struct {
-	Modules       map[string]AppModule
-	StoreDecoders sdk.StoreDecoderRegistry
+	Modules       []AppModuleSimulation    // array of app modules; we use an array for deterministic simulation tests
+	StoreDecoders sdk.StoreDecoderRegistry // functions to decode the key-value pairs from each module's store
+	ParamChanges  []simulation.ParamChange // list of parameters changes transactions run by the simulator
```

Comment:
> Looking through the refactor, it seems we only need this when we call `testAndRunTxs`. Instead, we can rather call a utility function that generates all the parameter changes: `generateParamChanges(sm, config.seed)`

`testAndRunTxs` contains the weighted operations for the simulation and is going to be refactored in #4869. I'd honestly leave that refactor for pt 3 since this PR is already bloated.

Reply:
Maybe we don't need to store this data as part of the state (just return it). Adding this function as part of the `SimulationManager` makes the expected simulation features clean and clear for users, as they won't need to look up for this external utility function to generate the param changes. 

> Instead, we can rather call a utility function that generates all the parameter changes: `generateParamChanges(sm, config.seed)`

Honestly, I think we should enforce apps to define their simulated parameter changes in each module. That was my previous reasoning as well for making the `AppModuleSimulation` mandatory. We shouldn't make the simulation an optional testing feature if we want to increase the security of the chains in the ecosystem.